connect-inject/health_check_handler.go

```diff
@@ -0,0 +1,317 @@
+// the consul health check accordingly. if the health check doesnt yet exist it will
+func (t *HealthCheckHandler) Reconcile() error {
+	podList, err := t.Clientset.CoreV1().Pods(corev1.NamespaceAll).List(ctx.Background(),
+		metav1.ListOptions{LabelSelector: labelInject})
+	if err != nil {
+		t.Log.Error("unable to get pods failing handler, err:%v", err)
+		return err
+	}
+	for _, pod := range podList.Items {
+		t.Log.Debug("processing Pod %v", pod.Name)
+		if pod.Status.Phase != corev1.PodRunning {
+			t.Log.Info("pod %v is not running", pod.Name, pod.Status.Phase)
+			continue
+		}
+		// 1. Fetch the health checks for this agent
+		err = t.updateConsulClient(&pod)
+		if err != nil {
+			t.Log.Error("unable to set client connection for %v", pod.Name)
+			continue
+		}
+		serviceID := t.getConsulServiceID(&pod)
+		healthCheckID := t.getConsulHealthCheckID(&pod)
+
+		// 2. Fetch the health checks on this agent related to this healthCheckID
+		filter := "Name == `" + healthCheckID + "`"
+		checks, err := t.Client.Agent().ChecksWithFilter(filter)
+		if err != nil {
+			t.Log.Error("unable to get agent health checks for %v", healthCheckID, filter)
+			continue
+		}
+		if checks == nil || len(checks) == 0 {
+			// TODO we need to handle this
+			t.Log.Error("======== we really shouldn't be here!, %v", checks)
+			continue
+		}
+		status, reason, err := t.getReadyStatusAndReason(&pod)
+		if err != nil {
+			t.Log.Error("unable to get pod status: %v", err)
+			continue
+		}
+		initialStatus := healthCheckPassing
+		if _, ok := checks[healthCheckID]; !ok {
+			// 3. If the check doesnt exist yet, create it!
+			if status == corev1.ConditionFalse {
+				initialStatus = healthCheckCritical
+			} else {
+				// no need to store a reason if passing
+				reason = ""
+			}
+			t.Log.Info("registering new health check for %v", pod.Name)
+			err = t.registerConsulHealthCheck(&pod, healthCheckID, serviceID, initialStatus, reason)
+			if err != nil {
+				t.Log.Error("unable to register health check %v", err)
+			}
+			continue
+		} else {
+			// 3. check does exist, update it if necessary
+			checkStatus := checks[healthCheckID].Status
+			fail := true
+			if checkStatus == healthCheckCritical && status == corev1.ConditionTrue {
+				// set to passing
+				fail = false
+				reason = ""
+			}
+			if (checkStatus == healthCheckPassing && status == corev1.ConditionFalse) || fail == false {
+				t.Log.Info("updating pod check for %v", pod.Name)
+				// Update the health check!
+				err = t.setConsulHealthCheckStatus(healthCheckID, reason, fail)
+				if err != nil {
+					t.Log.Error("unable to update health check status for %v", pod.Name)
+					continue
+				}
+			}
+		}
+		t.Log.Debug("no update required for %V", pod.Name)
+	}
+	return nil
+}
+
+func (t *HealthCheckHandler) getReadyStatusAndReason(pod *corev1.Pod) (status corev1.ConditionStatus, reason string, err error) {
+	for _, cond := range pod.Status.Conditions {
```

Comment:
If I run `kubectl describe`, I can see a more in-depth error, like:
```
Warning  Unhealthy  4s (x8 over 74s)  kubelet            Readiness probe failed: cat: /tmp/healthy: No such file or directory
```

This can be a "todo later" but it would be nice if we could use that as the reason.

Reply:
agreed, I've switched it to use a different metadata field but it does not have the whole readiness probe message, will add as a todo